File: python/rail402_x402/verify.py
"""On-chain verification of x402 payment proofs on Base via web3.py."""
from __future__ import annotations

import re
from typing import Optional

from web3 import Web3

from .types import NETWORKS, TRANSFER_TOPIC, Network, VerificationResult

_TX_HASH_RE = re.compile(r"^0x[0-9a-fA-F]{64}$")


def _topic_to_address(topic: str | bytes) -> str:
    hexstr = topic.hex() if isinstance(topic, (bytes, bytearray)) else topic
    hexstr = hexstr[2:] if hexstr.startswith("0x") else hexstr
    return Web3.to_checksum_address("0x" + hexstr[-40:])


def verify_payment(
    tx_hash: str,
    amount_atomic: str,
    pay_to: str,
    network: Network = "base",
    rpc_url: Optional[str] = None,
    min_confirmations: int = 1,
) -> VerificationResult:
    """Verify a USDC payment by inspecting its transaction receipt on Base.

    A proof is valid only when the transaction exists, succeeded, has enough
    confirmations, and contains a USDC ``Transfer`` to ``pay_to`` of at least
    ``amount_atomic``.
    """
    if not tx_hash or not _TX_HASH_RE.match(tx_hash):
        return VerificationResult(valid=False, error="Invalid or missing transaction hash")

    net = NETWORKS[network]
    w3 = Web3(Web3.HTTPProvider(rpc_url or net["default_rpc"]))

    try:
        receipt = w3.eth.get_transaction_receipt(tx_hash)
    except Exception:  # noqa: BLE001 - any RPC/lookup failure is "not found"
        return VerificationResult(valid=False, error="Transaction not found or not yet mined")

    if receipt.get("status") != 1:
        return VerificationResult(valid=False, error="Transaction reverted on-chain")

    if min_confirmations > 1:
        confirmations = w3.eth.block_number - receipt["blockNumber"] + 1
        if confirmations < min_confirmations:
            return VerificationResult(
                valid=False,
                error=f"Insufficient confirmations ({confirmations}/{min_confirmations})",
            )

    usdc = Web3.to_checksum_address(net["usdc"])
    expected_to = Web3.to_checksum_address(pay_to)
    required = int(amount_atomic)

    for log in receipt["logs"]:
        if Web3.to_checksum_address(log["address"]) != usdc:
            continue
        topics = log["topics"]
        if len(topics) < 3:
            continue
        topic0 = topics[0].hex() if hasattr(topics[0], "hex") else topics[0]
        if not topic0.lower().endswith(TRANSFER_TOPIC[2:]):
            continue
        if _topic_to_address(topics[2]) != expected_to:
            continue
        value = int(log["data"].hex() if hasattr(log["data"], "hex") else log["data"], 16)
        if value >= required:
            return VerificationResult(
                valid=True,
                tx_hash=tx_hash,
                payer_wallet=_topic_to_address(topics[1]),
                amount_atomic=str(value),
            )

    return VerificationResult(valid=False, error="No matching USDC transfer found in transaction")

File: src/verify.ts
import { createPublicClient, getAddress, http, type Hash } from "viem";
import { NETWORKS, TRANSFER_TOPIC } from "./constants";
import type {
  Address,
  TxHash,
  X402Network,
  X402PaymentProof,
  X402VerificationResult,
} from "./types";

/** A 32-byte topic encodes an address in its low 20 bytes. */
function topicToAddress(topic: string): Address {
  return getAddress(`0x${topic.slice(-40)}`);
}

export interface VerifyExpectation {
  /** Required amount in atomic USDC units. */
  amountAtomic: string;
  /** Wallet that must have received the funds. */
  payTo: Address;
  /** Network to verify against. Default "base". */
  network?: X402Network;
  /** Override RPC endpoint. */
  rpcUrl?: string;
  /** Minimum confirmations. Default 1. */
  minConfirmations?: number;
}

/**
 * Verify a payment proof by inspecting the on-chain transaction receipt on
 * Base. A proof is valid only when the referenced transaction:
 *   1. exists and succeeded,
 *   2. has the required number of confirmations, and
 *   3. contains a USDC `Transfer` to `payTo` of at least `amountAtomic`.
 *
 * Uses viem's JSON-RPC client; no indexer or third-party API required.
 */
export async function verifyPayment(
  proof: X402PaymentProof,
  expected: VerifyExpectation,
): Promise<X402VerificationResult> {
  if (!proof?.txHash || !/^0x[0-9a-fA-F]{64}$/.test(proof.txHash)) {
    return { valid: false, error: "Invalid or missing transaction hash" };
  }

  const network = expected.network ?? "base";
  const net = NETWORKS[network];
  const client = createPublicClient({ transport: http(expected.rpcUrl ?? net.defaultRpc) });

  let receipt;
  try {
    receipt = await client.getTransactionReceipt({ hash: proof.txHash as Hash });
  } catch {
    return { valid: false, error: "Transaction not found or not yet mined" };
  }

  if (receipt.status !== "success") {
    return { valid: false, error: "Transaction reverted on-chain" };
  }

  const minConfirmations = expected.minConfirmations ?? 1;
  if (minConfirmations > 1) {
    const head = await client.getBlockNumber();
    const confirmations = head - receipt.blockNumber + 1n;
    if (confirmations < BigInt(minConfirmations)) {
      return {
        valid: false,
        error: `Insufficient confirmations (${confirmations}/${minConfirmations})`,
      };
    }
  }

  const usdc = getAddress(net.usdc);
  const expectedTo = getAddress(expected.payTo);
  const requiredAmount = BigInt(expected.amountAtomic);

  for (const log of receipt.logs) {
    if (getAddress(log.address) !== usdc) continue;
    if (log.topics[0]?.toLowerCase() !== TRANSFER_TOPIC) continue;
    if (log.topics.length < 3) continue;

    const to = topicToAddress(log.topics[2] as string);
    if (to !== expectedTo) continue;

    const value = BigInt(log.data);
    if (value >= requiredAmount) {
      return {
        valid: true,
        txHash: proof.txHash as TxHash,
        payerWallet: topicToAddress(log.topics[1] as string),
        amountAtomic: value.toString(),
      };
    }
  }

  return { valid: false, error: "No matching USDC transfer found in transaction" };
}
